### PR TITLE
A couple of SIL fixes

### PR DESCRIPTION
**Pull request by slavapestov**
The ownership verifier got confused by code where a stored property of a class was initialized in-place with an enum value, or was the destination of a cast, erroneously flagging an ownership violation:

```
  %s = begin_borrow ...
  %e = ref_element_addr %s
  %d = init_enum_data_addr %e // or inject_enum_addr, unconditional_checked_cast_addr
```

DI got confused by code where more than one component of tuple value was initialized in-place via out parameters of an `apply` instruction, erroneously diagnosing that all but the first tuple component were uninitialized: 

```
  %t0 = tuple_element_addr %s, 0
  %t1 = tuple_element_addr %s, 1
  %f = function_ref ...
  apply %f(%t0, %t1)
```

I guess neither code pattern is generated anywhere in the ownership SIL pipeline today, but I imagine it will come up once we more aggressively eliminate `copy_addrs` before lowering ownership. Also, a change that I'm working on refactors some SILGen code to fix a bug, and the refactoring makes use of in-place initialization, which now generates the above code patterns instead of initializing a temporary and then performing a `copy_addr` from the temporary.

Finally, DI had some code to 'scalarize' certain tuple operations. After fixing the dataflow analysis to cope with instructions that have multiple non-load uses (such as an `apply` that initializes both components of a tuple), it appears that we no longer need this step at all.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 1]
```diff
@@ -543,16 +543,18 @@ bool InteriorPointerOperand::getImplicitUses(
...
     // Then handle users that we need to look at transitive uses of.
     if (Projection::isAddressProjection(user) ||
         isa<ProjectBlockStorageInst>(user) ||
         isa<OpenExistentialAddrInst>(user) ||
-        isa<InitExistentialAddrInst>(user) || isa<BeginAccessInst>(user) ||
-        isa<TailAddrInst>(user) || isa<IndexAddrInst>(user)) {
+        isa<InitExistentialAddrInst>(user) ||
+        isa<InitEnumDataAddrInst>(user) || isa<BeginAccessInst>(user) ||
+        isa<TailAddrInst>(user) || isa<IndexAddrInst>(user) ||
+        isa<UnconditionalCheckedCastAddrInst>(user)) {
```
Comment (gottesmm) [inline lib/SIL/Utils/OwnershipUtils.cpp:557]: @slavapestov This is incorrect. UnconditionalCheckedCastAddrInst does not have any results. Instead, you need to select its uses directly. Can you add an error test for unconditional_checked_cast_addr as well just to show via a test that we are correct here?

File: test/SILOptimizer/definite_init_tuple.sil [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+sil [ossa] @$s19definite_init_tuple1SV1xx_xtvpfi : $@convention(thin) <T> () -> (@out T, @out T)
+
+sil [ossa] @$s19definite_init_tuple1SVACyxGycfC : $@convention(method) <T> (@thin S<T>.Type) -> @out S<T> {
+bb0(%0 : $*S<T>, %1 : $@thin S<T>.Type):
+  %2 = alloc_box $<τ_0_0> { var S<τ_0_0> } <T>, var, name "self"
+  %3 = mark_uninitialized [rootself] %2 : $<τ_0_0> { var S<τ_0_0> } <T>
+  %4 = project_box %3 : $<τ_0_0> { var S<τ_0_0> } <T>, 0
+  %5 = struct_element_addr %4 : $*S<T>, #S.x
```
Comment (gottesmm) [inline test/SILOptimizer/definite_init_tuple.sil:18]: @slavapestov This has far fewer tests than the old version that was deleted above. Are you sure we didn't lose any code coverage?

**Pull request merged**